cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/PreviewRuntimeService.java

```diff
@@ -0,0 +1,92 @@
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.logging.LoggingContextAccessor;
+import co.cask.cdap.common.logging.ServiceLoggingContext;
+import co.cask.cdap.data2.datafabric.dataset.service.DatasetService;
+import co.cask.cdap.internal.app.namespace.DefaultNamespaceEnsurer;
+import co.cask.cdap.internal.app.runtime.artifact.SystemArtifactLoader;
+import co.cask.cdap.internal.app.services.ApplicationLifecycleService;
+import co.cask.cdap.internal.app.services.ProgramLifecycleService;
+import co.cask.cdap.logging.appender.LogAppenderInitializer;
+import co.cask.cdap.proto.id.NamespaceId;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.common.util.concurrent.Futures;
+import com.google.inject.Inject;
+
+/**
+ * Service that runs the preview.
+ */
+public class PreviewRuntimeService extends AbstractIdleService {
+
+  private final DatasetService datasetService;
+  private final LogAppenderInitializer logAppenderInitializer;
+  private final ApplicationLifecycleService applicationLifecycleService;
+  private final SystemArtifactLoader systemArtifactLoader;
+  private final ProgramRuntimeService programRuntimeService;
+  private final ProgramLifecycleService programLifecycleService;
+  private final DefaultNamespaceEnsurer namespaceEnsurer;
+
+  @Inject
+  PreviewRuntimeService(DatasetService datasetService, LogAppenderInitializer logAppenderInitializer,
+                        ApplicationLifecycleService applicationLifecycleService,
+                        SystemArtifactLoader systemArtifactLoader, ProgramRuntimeService programRuntimeService,
+                        ProgramLifecycleService programLifecycleService, DefaultNamespaceEnsurer namespaceEnsurer) {
+    this.datasetService = datasetService;
+    this.logAppenderInitializer = logAppenderInitializer;
+    this.applicationLifecycleService = applicationLifecycleService;
+    this.systemArtifactLoader = systemArtifactLoader;
+    this.programRuntimeService = programRuntimeService;
+    this.programLifecycleService = programLifecycleService;
+    this.namespaceEnsurer = namespaceEnsurer;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    datasetService.startAndWait();
+
+    // It is recommended to initialize log appender after datasetService is started,
+    // since log appender instantiates a dataset.
+    logAppenderInitializer.initialize();
+
+    LoggingContextAccessor.setLoggingContext(new ServiceLoggingContext(NamespaceId.SYSTEM.getNamespace(),
+                                                                       Constants.Logging.COMPONENT_NAME,
+                                                                       Constants.Service.PREVIEW_HTTP));
+    Futures.allAsList(
+      applicationLifecycleService.start(),
+      systemArtifactLoader.start(),
+      programRuntimeService.start(),
+      programLifecycleService.start()
+    ).get();
+
+    namespaceEnsurer.startAndWait();
```

Comment:
shouldn't matter, but perhaps just add this in the futures list above as well - doesn't seem like there is any reason why it should be done at the end.